src/tree_data.c

```diff
@@ -1653,7 +1653,28 @@ lyd_make_canonical(const struct lys_node *schema, const char *val_str, int val_s
     }
 
     /* parse the value into a fake leaf */
-    node = lyd_create_leaf(schema, str, 0, 0);
+    struct lyd_node_leaf_list *node_leaf_list;
+
+    node_leaf_list = calloc(1, sizeof *node_leaf_list);
+    LY_CHECK_ERR_RETURN(!node_leaf_list, LOGMEM(schema->module->ctx), NULL);
+
+    node_leaf_list->schema = (struct lys_node *)schema;
+    node_leaf_list->validity = ly_new_node_validity(schema);
+    if (resolve_applies_when(schema, 0, NULL)) {
+        node_leaf_list->when_status = LYD_WHEN;
+    }
+    node_leaf_list->prev = (struct lyd_node *)node_leaf_list;
+    node_leaf_list->value_type = ((struct lys_node_leaf *)schema)->type.base;
+    node_leaf_list->value_str = strdup(str);
+    node_leaf_list->dflt = 0;
+
+    if (!lyp_parse_value(&((struct lys_node_leaf *)schema)->type, &node_leaf_list->value_str, NULL, node_leaf_list, NULL, NULL, 0, 0)) {
```

Comment:
Now that the code can be tailored to the specific use-case, please do so. `lyp_parse_value` does not need `validity` or `when_status` set, so do not bother with that (some minor performance gain).